README.rst
==============
Libcast XBlock
==============

.. image:: https://circleci.com/gh/openfun/libcast-xblock/tree/master.svg?style=svg
    :target: https://circleci.com/gh/openfun/libcast-xblock/tree/master

This xblock is used by France University Numérique to migrate from one video providers to another.

We recommend to use Youtube xblock included in edx-platform

Install
=======

::

    pip install --extra-index-url https://pypi.fury.io/openfun libcast-xblock


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.6.1] - 2020-03-23

### Removed

- Revert commit which was embeding Video-JS player as it breaks build
  and should not be necessary.

## [0.6.0] - 2020-02-13

### Added

- Implement experimental support for BokeCC video provider (legacy feature for
  developped 3 years ago for a white brand) See openfun/libcast-xblock#8

## [0.5.0] - 2018-03-21

### Added

- Add a CHANGELOG (yay)
- Add packaging good practices: all package configuration stands in the
  `setup.cfg` file
- Automate the packaging workflow with CircleCI

### Changed

- Update installation instructions in the README


libcast_xblock/libcast_xblock.py
"""XBlock for videos stored by Youtube."""

import logging
import random
import string

import pkg_resources

from django.conf import settings as django_settings
from django.template import Context, Template
from django.contrib.staticfiles.storage import staticfiles_storage

# TODO actually translate the app
from django.utils.translation import ugettext_lazy

# from django.utils.translation import ugettext as _
import webob

from xblock.core import XBlock
from xblock.fields import Boolean, Scope, String
from xblock.fragment import Fragment
from xblockutils.studio_editable import StudioEditableXBlockMixin


logger = logging.getLogger(__name__)


@XBlock.needs("settings")
class LibcastXBlock(StudioEditableXBlockMixin, XBlock):
    """
    Play videos based on a modified videojs player. This XBlock supports
    subtitles and multiple resolutions.
    """

    display_name = String(
        help=ugettext_lazy(
            "The name students see. This name appears in "
            "the course ribbon and as a header for the video."
        ),
        display_name=ugettext_lazy("Component Display Name"),
        default=ugettext_lazy("New video"),
        scope=Scope.settings,
    )

    video_id = String(
        scope=Scope.settings,
        help=ugettext_lazy(
            "Fill this with the ID of the video found in the video uploads dashboard"
        ),
        default="",
        display_name=ugettext_lazy("Video ID"),
    )

    is_youtube_video = Boolean(
        help=ugettext_lazy("Is this video stored on Youtube?"),
        display_name=ugettext_lazy("Video storage"),
        scope=Scope.settings,
        default=False,  # Stored on Videofront by default
    )

    is_bokecc_video = Boolean(
        help=ugettext_lazy("Is this video stored on Bokecc?"),
        display_name=ugettext_lazy("Video storage"),
        scope=Scope.settings,
        default=False,  # Stored on Videofront by default
    )

    allow_download = Boolean(
        help=ugettext_lazy("Allow students to download this video."),
        display_name=ugettext_lazy("Video download allowed"),
        scope=Scope.settings,
        default=True,
    )

    adways_id = String(
        scope=Scope.settings,
        help=ugettext_lazy(""),
        default="",
        display_name=ugettext_lazy("Adways ID"),
    )

    @property
    def editable_fields(self):
        fields = (
            "display_name",
            "video_id",
            "is_youtube_video",
            "is_bokecc_video",
            "allow_download",
        )
        adways_courses = getattr(django_settings, "ENABLE_ADWAYS_FOR_COURSES", [])
        if self.course_key_string in adways_courses:
            fields += ("adways_id",)
        return fields

    def __init__(self, *args, **kwargs):
        super(LibcastXBlock, self).__init__(*args, **kwargs)

    @property
    def course_key_string(self):
        try:
            return unicode(self.location.course_key)
        except AttributeError:
            return ""

    @property
    def resource_slug(self):
        return None if self.video_id is None else self.video_id.strip()

    def transcript_root_url(self):
        url = self.runtime.handler_url(self, "transcript")
        # url is suffixed with '?' in preview mode
        return url.strip("?")

    @XBlock.handler
    def transcript(self, request, dispatch):
        """
        Proxy view for downloading subtitle files

        <track> elements that point to external resources are not supported by
        IE11 and Microsoft Edge:
        http://stackoverflow.com/questions/35138642/ms-edge-video-cross-origin-subtitles-fail
        As a consequence, we need to download the subtitle file server-side.
        """
        from videoproviders.subtitles import get_vtt_content

        subtitle_url = request.GET.get("url")
        if not subtitle_url:
            return webob.Response(status=404)
        caps = get_vtt_content(subtitle_url) or ""
        return webob.Response(caps, content_type="text/vtt")

    def get_icon_class(self):
        """CSS class to be used in courseware sequence list."""
        return "video"

    def is_studio(self):
        studio = False
        try:
            studio = self.runtime.is_author_mode
        except AttributeError:
            pass
        return studio

    def student_view(self, context=None):
        fragment = Fragment()
        if self.is_youtube_video and self.video_id:
            self.get_youtube_content(fragment)
        if self.is_bokecc_video and self.video_id:
            self.get_bokecc_content(fragment)
        else:
            self.get_videofront_content(fragment)
        return fragment

    def get_videofront_content(self, fragment):
        from videoproviders.api import videofront

        template_content = self.resource_string("public/html/videofront.html")
        template = Template(template_content)
        messages = []  # tuple list
        context = {
            "display_name": self.display_name,
            "video_id": self.resource_slug,
            "messages": messages,
            "allow_download": self.allow_download,
            "downloads": [],
            "sources": [],
            "subtitles": [],
            "thumbnail_url": "",
            "transcript_root_url": self.transcript_root_url(),
        }
        if self.resource_slug:
            try:
                videofront_client = videofront.Client(self.course_key_string)
                video = videofront_client.get_video_with_subtitles(self.resource_slug)
                context["subtitles"] = video["subtitles"]
                context["thumbnail_url"] = video["thumbnail_url"]

                # Sort video sources by decreasing resolution
                video_sources = sorted(video["video_sources"], key=lambda s: -s["res"])

                # Streaming sources
                source_labels = {
                    "HD": "1080p",
                    "SD": "720p",
                    "LD": "480p",
                    "UL": "240p",
                }
                context["sources"] = [
                    {
                        "url": source["url"],
                        "label": source_labels.get(source["label"], source["label"]),
                        "res": source["res"],
                    }
                    for source in video_sources
                ]

                # Download links
                download_labels = {
                    "HD": "Haute (1080p)",
                    "SD": "Normale (720p)",
                    "LD": "Mobile (480p)",
                    "UL": "Basse (240p)",
                }
                context["downloads"] = [
                    {
                        "url": source["url"],
                        "label": download_labels.get(source["label"], source["label"]),
                    }
                    for source in video_sources
                ]

            except videofront.MissingCredentials as e:
                messages.append(("error", e.verbose_message))
            except videofront.ClientError as e:
                # Note that we may not log an exception here, because unicode
                # messages cannot be encoded by the logger
                logger.error(e.message)
                messages.append(
                    ("error", ugettext_lazy("An unexpected error occurred."))
                )
        else:
            messages.append(
                (
                    "warning",
                    ugettext_lazy(
                        "You need to define a valid video ID. "
                        "Video IDs for your course can be found in the video upload"
                        " dashboard."
                    ),
                )
            )

        content = template.render(Context(context))
        fragment.add_content(content)

        fragment.add_css(self.resource_string("public/css/style.css"))
        # Load video-js stylesheet from FUN; if this becomes a problem, we can
        # always copy the stylesheet to the xblock.
        fragment.add_css_url(staticfiles_storage.url("fun/js/vendor/videojs/video-js.min.css"))
        fragment.add_javascript(self.resource_string("public/js/videofront.js"))

        fragment.initialize_js(
            "VideoXBlock",
            json_args={
                "course_id": self.course_key_string,
                "video_id": self.resource_slug,
                "adways_id": self.adways_id,
                "block_id": self.location.block_id,
            },
        )

    def get_youtube_content(self, fragment):
        # iframe element id
        element_id = "".join(
            [random.choice(string.ascii_lowercase) for _ in range(0, 20)]
        )

        # Add html code
        template_content = self.resource_string("public/html/youtube.html")
        template = Template(template_content)
        context = {
            "display_name": self.display_name,
            "video_id": self.resource_slug,
            "element_id": element_id,
        }
        content = template.render(Context(context))
        fragment.add_content(content)

        # Add youtube event logger
        fragment.add_javascript(self.resource_string("public/js/youtube.js"))
        fragment.initialize_js(
            "YoutubePlayer",
            json_args={
                "course_id": self.course_key_string,
                "video_id": self.resource_slug,
                "element_id": element_id,
            },
        )

    def get_bokecc_content(self, fragment):
        from videoproviders.api import bokecc

        # iframe element id
        element_id = "".join(
            [random.choice(string.ascii_lowercase) for _ in range(0, 20)]
        )
        # This part will need to be moved into fun-apps as a VideoClient
        bokecc_client = bokecc.Client(self.course_key_string)
        video = bokecc_client.get_video(self.resource_slug)

        # Add html code
        template_content = self.resource_string("public/html/bokecc.html")
        template = Template(template_content)
        context = {
            "display_name": self.display_name,
            "video_id": self.resource_slug,
            "js_script_url": video["js_script_url"],
            "element_id": element_id,
        }
        content = template.render(Context(context))
        fragment.add_content(content)

        # Add bokeecc event logger
        fragment.add_javascript(self.resource_string("public/js/bokecc.js"))
        fragment.initialize_js(
            "BokeccPlayer",
            json_args={
                "course_id": self.course_key_string,
                "video_id": self.resource_slug,
                "element_id": element_id,
            },
        )

    def resource_string(self, path):
        """Handy helper for getting resources from our kit."""
        data = pkg_resources.resource_string(__name__, path)
        return data.decode("utf8")

    @staticmethod
    def workbench_scenarios():
        """
        Define default workbench scenarios
        """
        return [
            (
                "A little Video",
                """
                    <libcast_xblock video_id="06C1C538BF97169B9C33DC5901307461" xblock-family="xblock.v1" is_bokecc_video="true"/>
             """,
            )
        ]


libcast_xblock/public/js/bokecc.js
function BokeccPlayer(runtime, element, args) {
    'use strict';
    // Event logging
    function log(eventName, data) {
        var logInfo = {
            course_id: args.course_id,
            video_id: args.video_id,
        };
        if (data) {
            $.extend(logInfo, data);
        }
        Logger.log(eventName, logInfo);
    }

    function bookeecc_getSWF(swfID) {
        if (window.document[swfID]) {
            return window.document[swfID];
        } else if (navigator.appName.indexOf("Microsoft") == -1) {
            if (document.embeds && document.embeds[swfID]) {
                return document.embeds[swfID];
            }
        } else {
            return document.getElementById(swfID);
        }
    }
    // The on_cc_player_init is a callback from the flash plugin
    if (typeof window.on_cc_player_init == "undefined") {
        window.on_cc_player_init = function (vid, objectId) {
            var config = {};
            // define global event functions

            window['bokecc_on_ps_'+objectId] = function() {
                var player = bookeecc_getSWF(objectId);
                log('play_video',player.getPosition());
            }
            window['bokecc_on_sk_'+objectId] = function() {
                var player = bookeecc_getSWF(objectId);
                log('seek_video','new_time');
            }
            window['bokecc_on_sv_'+objectId] = function() {
                var player = bookeecc_getSWF(objectId);
                log('stop_video',player.getPosition());
            }
            window['bokecc_on_pv_'+objectId] = function() {
                var player = bookeecc_getSWF(objectId);
                log('pause_video',player.getPosition());
            }
            config.rightmenu_enable = 0;
            config.on_player_start = 'bokecc_on_ps_'+objectId;
            config.on_player_seek = 'bokecc_on_sk_'+objectId;
            config.on_player_stop = 'bokecc_on_sv_'+objectId;
            config.on_player_pause = 'bokecc_on_pv_'+objectId;
            config.on_player_resume = 'bokecc_on_ps_'+objectId;
            var player = bookeecc_getSWF(objectId)
            player.setConfig(config);
        }
    }

}


libcast_xblock/public/js/youtube.js
var player;
function YoutubePlayer(runtime, element, args) {
    'use strict';

    // Insert youtube script
    // We need to do this even if the script was already inserted, in order to
    // handle multiple video xblocks in the same page.
    var tag = document.createElement('script');
    tag.src = "https://www.youtube.com/iframe_api";
    var firstScriptTag = document.getElementsByTagName('script')[0];
    firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);

    // This function gets called after the youtube script was inserted
    window.onYouTubeIframeAPIReady = function() {
        var onPlayerReady = function(e) {
            log('video_player_ready');
        };

        var onPlayerStateChanged = function(e) {
            // Note that there is no event for seeking in the video
            if (e.data == YT.PlayerState.UNSTARTED) {
                log('load_video');
            } else if (e.data == YT.PlayerState.ENDED) {
                log('stop_video');
            } else if (e.data == YT.PlayerState.PLAYING) {
                log('play_video', {currentTime: player.getCurrentTime()});
            } else if (e.data == YT.PlayerState.PAUSED) {
                log('pause_video', {currentTime: player.getCurrentTime()});
            } else if (e.data == YT.PlayerState.BUFFERING) {
            } else if (e.data == YT.PlayerState.CUED) {
            }
        };

        player = new YT.Player(args.element_id, {
           events: {
               'onReady': onPlayerReady,
               'onStateChange': onPlayerStateChanged
           }
        });
    };

    // Event logging
    function log(eventName, data) {
        var logInfo = {
          course_id: args.course_id,
          video_id: args.video_id,
        };
        if (data) {
          $.extend(logInfo, data);
        }
        Logger.log(eventName, logInfo);
    }
}


libcast_xblock/public/js/videofront.js
function VideoXBlock(runtime, element, args) {
  'use strict';
  var require = require || RequireJS.require;
  require(['videoplayer-fun', 'adways-player'], function(videoplayer, adways) {

    var videoPlayerElement = $(element).find('.videoplayer');
    var transcriptElement = videoPlayerElement.find(".transcript");
    var video = videoPlayerElement.find('video')[0]

    // disable video element contextual menu...
    $(video).bind('contextmenu',function() { return false; });

    var player = videoplayer(video);

    if (args.adways_id) {
      var experience = new adways.interactive.Experience();
      experience.setPublicationID(args.adways_id);
      experience.setPlayerAPI(player);
      experience.setPlayerClass("videojs");
      experience.setSceneTimeReference(false); //true for live interactivity (i.e. based on user time instead of video time).
      experience.load();
    }

    // Configure transcripts
    player.one('loadedmetadata', function() {
      var tracks = player.textTracks();

      // Change track
      tracks.addEventListener('change', function() {

        var enableTranscript = false;
        for (var t = 0; t < this.length; t++) {
          var track = this[t];
          if (track.mode === 'showing') {
            showTranscript(track);
            enableTranscript = true;
          }
        }
        if (!enableTranscript) {
          disableTranscript();
        }
      });

      // Highlight current cue
      for (var t = 0; t < tracks.length; t++) {
        tracks[t].addEventListener('cuechange', oncuechange);
      }
    });

    var showTranscript = function(track) {
      var cues = track.cues;

      // We need to check whether the track is still the one currently showing.
      if (track.mode !== "showing") {
        return;
      }

      // Cues may not be loaded yet. If not, wait until they are. This is
      // suboptimal, but there is no other event to help us determine whether a
      // track was correctly loaded.
      if (!cues || cues.length === 0) {
        window.setTimeout(function() { showTranscript(track); }, 2);
      }

      var htmlContent = "";
      for (var c = 0; c < cues.length; c++) {
        var cue = cues[c];
        htmlContent += "<span class='cue' begin='" + cue.startTime + "'>&nbsp;-&nbsp;" + cue.text + "</span><br/>\n";
      }

      player.width("61%");
      videoPlayerElement.addClass("transcript-enabled");
      transcriptElement.html(htmlContent);

      // Go to time on cue click
      transcriptElement.find(".cue").click(function() {
          player.currentTime($(this).attr('begin'));
      });
    };

    var disableTranscript = function() {
      videoPlayerElement.removeClass("transcript-enabled");
      player.width("100%");
    };

    var oncuechange = function() {
      transcriptElement.find(".current.cue").removeClass("current");
      var cueElement;
      for (var c = 0; c < this.activeCues.length; c++) {
        cueElement = transcriptElement.find(".cue[begin='" + this.activeCues[c].startTime + "']");
        cueElement.addClass("current");
      }
      if (cueElement) {
        // Scroll to cue
        var newtop = transcriptElement.scrollTop() - transcriptElement.offset().top + cueElement.offset().top;
        transcriptElement.animate({
            scrollTop: newtop
        }, 500);
      }
    };

    // Listen to events
    var logTimeOnEvent = function(eventName, logEventName, currentTimeKey, data) {
      player.on(eventName, function() {
        logTime(logEventName, data, currentTimeKey);
      });
    };
    var logTime = function(logEventName, data, currentTimeKey) {
      data = data || {};
      currentTimeKey = currentTimeKey || 'currentTime';
      data[currentTimeKey] = parseInt(player.currentTime());
      log(logEventName, data);
    };
    var logOnEvent = function(eventName, logEventName, data) {
      data = data || {};
      player.on(eventName, function() {
          log(logEventName, data);
      });
    };

    function log(eventName, data) {
        var logInfo = _.extend({
            // Old FUN params
            course_id: args.course_id, // this value is now useless
            video_id: args.video_id,   // this field is now useless

            // edX params
            id: args.block_id,         // xblock id in course
            code: args.video_id        // video ID
        }, data);
        Logger.log(eventName, logInfo);
    };


    logTimeOnEvent('seeked', 'seek_video', 'new_time');
    logTimeOnEvent('ended', 'stop_video');
    logTimeOnEvent('pause', 'pause_video');
    logTimeOnEvent('play', 'play_video');
    logOnEvent('loadstart', 'load_video');
    log('video_player_ready');
    // Note that we have no show/hide transcript button, so there is nothing to
    // log for these events

    player.on('ratechange', function() {
      logTime('speed_change_video', { newSpeed: player.playbackRate() });
    });
  });
}


libcast_xblock/public/js/libcast_xblock.js
function LibcastXBlock(runtime, element, args) {
  'use strict';
  var require = require || RequireJS.require;
  require(['libcast'], function(libcast) {

    var videoPlayerElement = $(element).find('.videoplayer');
    var transcriptElement = videoPlayerElement.find(".transcript");
    var player = libcast(videoPlayerElement.find('video')[0]);

    // Configure transcripts
    player.one('loadedmetadata', function() {
      var tracks = player.textTracks();

      // Change track
      tracks.addEventListener('change', function() {

        var enableTranscript = false;
        for (var t = 0; t < this.length; t++) {
          var track = this[t];
          if (track.mode === 'showing') {
            showTranscript(track);
            enableTranscript = true;
          }
        }
        if (!enableTranscript) {
          disableTranscript();
        }
      });

      // Highlight current cue
      for (var t = 0; t < tracks.length; t++) {
        tracks[t].addEventListener('cuechange', oncuechange);
      }
    });

    var showTranscript = function(track) {
      var cues = track.cues;

      // We need to check whether the track is still the one currently showing.
      if (track.mode !== "showing") {
        return;
      }

      // Cues may not be loaded yet. If not, wait until they are. This is
      // suboptimal, but there is no other event to help us determine whether a
      // track was correctly loaded.
      if (!cues || cues.length === 0) {
        window.setTimeout(function() { showTranscript(track); }, 2);
      }

      var htmlContent = "";
      for (var c = 0; c < cues.length; c++) {
        var cue = cues[c];
        htmlContent += "<span class='cue' begin='" + cue.startTime + "'>&nbsp;-&nbsp;" + cue.text + "</span><br/>\n";
      }

      player.width("61%");
      videoPlayerElement.addClass("transcript-enabled");
      transcriptElement.html(htmlContent);

      // Go to time on cue click
      transcriptElement.find(".cue").click(function() {
          player.currentTime($(this).attr('begin'));
      });
    };

    var disableTranscript = function() {
      videoPlayerElement.removeClass("transcript-enabled");
      player.width("100%");
    };

    var oncuechange = function() {
      transcriptElement.find(".current.cue").removeClass("current");
      var cueElement;
      for (var c = 0; c < this.activeCues.length; c++) {
        cueElement = transcriptElement.find(".cue[begin='" + this.activeCues[c].startTime + "']");
        cueElement.addClass("current");
      }
      if (cueElement) {
        // Scroll to cue
        var newtop = transcriptElement.scrollTop() - transcriptElement.offset().top + cueElement.offset().top;
        transcriptElement.animate({
            scrollTop: newtop
        }, 500);
      }
    };

    // Listen to events
    var logTimeOnEvent = function(eventName, logEventName, currentTimeKey, data) {
      player.on(eventName, function() {
        logTime(logEventName, data, currentTimeKey);
      });
    };
    var logTime = function(logEventName, data, currentTimeKey) {
      data = data || {};
      currentTimeKey = currentTimeKey || 'currentTime';
      data[currentTimeKey] = parseInt(player.currentTime());
      log(logEventName, data);
    };
    var logOnEvent = function(eventName, logEventName, data) {
      data = data || {};
      player.on(eventName, function() {
          log(logEventName, data);
      });
    };
    function log(eventName, data) {
        var logInfo = {
          course_id: args.course_id,
          video_id: args.video_id,
        };
        if (data) {
          $.extend(logInfo, data);
        }
        Logger.log(eventName, logInfo);
    }

    logTimeOnEvent('seeked', 'seek_video', 'new_time');
    logTimeOnEvent('ended', 'stop_video');
    logTimeOnEvent('pause', 'pause_video');
    logTimeOnEvent('play', 'play_video');
    logOnEvent('loadstart', 'load_video');
    log('video_player_ready');
    // Note that we have no show/hide transcript button, so there is nothing to
    // log for these events

    player.on('ratechange', function() {
      logTime('speed_change_video', { newSpeed: player.playbackRate() });
    });
  });
}
